### PR TITLE
Add Symfony/Validator articles

### PR DESCRIPTION
**Pull request by enumag**

### CODE REVIEW

File: source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+tested: true
+test_slug: SymfonyValidatorComparisonConstraints
+---
+
+
+Use cases
+----
+
+- For date interval ensure that starting date is lower than ending date.
+- Password confirmation value should be the same as password.
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md]: It would stick only to the date, as it is mentioned in the example bellow.

File: source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+The correct solution
+----
+
+The point of this article is that Symfony actually does have an [easy mechanism](http://symfony.com/doc/current/reference/constraints/Expression.html) to do these validations. The Expression constraint is not something new either. It's just not covered in the documentation very well and less experienced developers can easily miss it.
+
+The Expression constraint takes advantage of the [ExpressionLanguage](http://symfony.com/doc/current/components/expression_language.html) component. In the expression you can use the `value` and `this` placeholders to access the value itself and the underlying object respectively.
+
+
+Example
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md]: Could you please use ### for headlines? It's convetional here

File: source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+     * @Assert\Expression("value >= this.startDate")
+     */
+    protected $endDate;
+}
+```
+
+Indeed it's this easy! More importantly the Expression constraint can help you solve many other situations. Plus the ExpressionLanguage can be [extended](http://symfony.com/doc/current/components/expression_language/extending.html) with your own functions.
+
+
+Usage with Nette Framework
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md]: This could be in standalone article. I think it is not related to the expression language itself, rather implementation of Symfony Validator in Nette.

File: source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+----
+
+If you want to use the Symfony/Validator component and the Expression constraint in your Nette application you will need these libraries:
+
+- [Kdyby/Validator](https://github.com/Kdyby/Validator)
+- [Arachne/ExpressionLanguage](https://github.com/Arachne/ExpressionLanguage)
+- [Arachne/Doctrine](https://github.com/Arachne/Doctrine) (optional, enables caching for the ExpressionLanguage parser)
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md]: As for the article, try to highlight top 3 important things with **bold**

File: source/_posts/2017/2017-01-28-symfony-validator-conditional-constraints.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    /**
+     * @var string
+     * @Assert\NotBlank(groups = {"person"})
+     */
+    protected $lastname;
+}
+```
+
+Next you need to determine which group to validate. You can do that manually or you can use one of the advanced features of Symfony/Validator called GroupSequenceProviderInterface.
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-28-symfony-validator-conditional-constraints.md]: Here could be some headline? Or maybe number for step?

File: source/_posts/2017/2017-01-28-symfony-validator-conditional-constraints.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    }
+}
+```
+
+And that's it! Now Validator will validate the client differently based on the `$type` property and give you all violations.
+
+In the end conditional validation is surprisingly easy with Symfony/Validator. The solution is just a bit hiddden and not widely known.
+
+
+Usage with Nette Framework
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-28-symfony-validator-conditional-constraints.md]: Again, I would put this to standalone article.

File: tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php [hunk 7]
```diff
@@ -105,4 +105,10 @@ protected function createPresenter(): IPresenter
 
         return $categoryPresenter;
     }
+
+
+    private function readFile(string $file): string
+    {
+        return str_replace("\r\n", "\n", file_get_contents($file));
```
Comment (TomasVotruba) [inline tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php]: What this exactly do?

File: tests/bootstrap.php [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+use Doctrine\Common\Annotations\AnnotationRegistry;
+
+require __DIR__ . '/../vendor/autoload.php';
+
+AnnotationRegistry::registerLoader('class_exists');
```
Comment (TomasVotruba) [inline tests/bootstrap.php]: Isn't there a better way to load this without overloading it here?

File: tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php (hunk 7 above)
Comment (enumag) [inline tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php]: The test failed on windows because of different newlines.

File: tests/bootstrap.php (hunk 8 above)
Comment (enumag) [inline tests/bootstrap.php]: I don't know. I always did it this way.

File: tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php (hunk 7 above)
Comment (TomasVotruba) [inline tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php]: Could include this to name of the method? So it's explicit

File: tests/bootstrap.php (hunk 8 above)
Comment (TomasVotruba) [inline tests/bootstrap.php]: I Googled a lot and it seems there isn't. Alright then

File: source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md (hunk 3 above)
Comment (enumag) [inline source/_posts/2017/2017-01-27-symfony-validator-comparison-constraints.md]: I don't intend to write a separate article about Kdyby/Validator so I'l like to keep the note here.

**Comment by enumag** [comment]
@TomasVotruba Thanks for the review! :-)

File: tests/Posts/Year2017/ListeningNetteComponents/Component/BasketContentControl/BasketContentControl.php [hunk 9]
```diff
@@ -33,10 +34,13 @@ public function onProductAddedToBasketEvent(ProductAddedToBasketEvent $productAd
...
-        $this->template->setParameters([
+        /** @var Template $template */
+        $template = $this->getTemplate();
+
+        $template->setParameters([
             'products' => $this->products
         ]);
 
-        $this->template->render(__DIR__ . '/templates/default.latte');
+        $template->render(__DIR__ . '/templates/default.latte');
```
Comment (TomasVotruba) [inline tests/Posts/Year2017/ListeningNetteComponents/Component/BasketContentControl/BasketContentControl.php]: Thanks for the improvement!

You can  take it one step further (will do while merging) to using 2nd argument of `render()` method:

```php
$template->render(__DIR__ . '/templates/default.latte', [
    'products' => $this->products
]);
```

File: tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php [hunk 10]
```diff
@@ -57,6 +58,8 @@ public function __construct(
         BasketContentControlFactoryInterface $basketContentControlFactory,
         EventDispatcherInterface $eventDispatcher
     ) {
+        parent::__construct();
```
Comment (TomasVotruba) [inline tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php]: Why is this needed?

File: tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php [hunk 11]
```diff
@@ -78,7 +81,9 @@ public function startup()
 
     public function renderDefault()
     {
-        $this->template->setParameters([
+        /** @var Template $template */
+        $template = $this->getTemplate();
```
Comment (TomasVotruba) [inline tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php]: I didn't know about the magic here. Thanks for the pointing that out.

File: .travis.yml [hunk 12]
```diff
@@ -28,7 +28,11 @@ script:
     # run tests for tested posts (with "tested: true" in header configuration)
     - vendor/bin/phpunit $PHPUNIT_FLAGS
     # check coding standard (defined in composer.json "scripts" section)
-    - if [[ "$CHECK_CS" != "" ]]; then composer check-cs; fi
+    - |
+        if [[ "$CHECK_CS" != "" ]]; then
```
Comment (TomasVotruba) [inline .travis.yml]: I would split those, since coding standard and static analytics are different.

See: https://github.com/Symplify/Symplify/blob/e1c4a545ee3ad5f46c0f2e1a4c260602a6f67530/.travis.yml#L6

File: composer.json [hunk 13]
```diff
@@ -33,12 +34,16 @@
     },
     "scripts": {
         "check-cs": [
-            "vendor/symplify/coding-standard/bin/symplify-cs check src",
-            "vendor/symplify/coding-standard/bin/symplify-cs check tests"
+            "vendor/bin/symplify-cs check src",
+            "vendor/bin/symplify-cs check tests"
```
Comment (TomasVotruba) [inline composer.json]: Thank you. This probably some debug leftover.

File: composer.json [hunk 14]
```diff
@@ -5,21 +5,21 @@
...
-        "doctrine/annotations": "^1.3",
-        "doctrine/cache": "^1.6",
-        "sllh/iso-codes-validator": "^3.0",
         "symfony/expression-language": "^3.1",
         "symfony/intl": "^3.1",
-        "symfony/validator": "^3.2"
+        "symfony/validator": "^3.2",
+
+        "phpunit/phpunit": "^5.7",
+        "symplify/coding-standard": "^1.3"
```
Comment (TomasVotruba) [inline composer.json]: Thanks. Is there some script to do that automatically?

**Comment by TomasVotruba** [comment]
@enumag Happy to comment :+1:

File: tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php (hunk 10 above)
Comment (enumag) [inline tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php]: PHPStan complains if parent is not called. An in case of presenter it seems better to call it although it's not needed in this case.

File: .travis.yml [hunk 15]
```diff
@@ -21,14 +21,18 @@ env:
         - secure: XlpJL/8eke9LX831DFKDi2CHomqxZxXM1+Z4QY6AoPmiq0CHpBSYk62GbNooSjDHwsTjEeDBHEextZPWtMhXbTJG71V7eYXY2p3NfSuYVjUTBPy1sfiJX2anYv1uV8WU8tmQqF3aIQ/tYLLRiBIYdkskewpFYdcdvht5E3baajcVneGqKXAIXrYMg7zkpBoNvDG2X0ex2LyFjDiqIdBZM8CUioCVt6IxovhtsyiQicuAovghCN7lmCWh/Klj8lDKi1FzjZ6xYZoWzrzbguLJRqxxrJXbx1f2JZAYixiRE3vOIyJLWnDgv2/GRlFbsCuuWGF55P4oSeXGArButYHJz3GDWCX+CAlgsBvGcFAdwrK5T3hNo7jRTzjnCSaXHiN/dUUFTAyFvqvAJQSBg7OoUpw5yJSOO9Yu1IMMyoUkzfWr5ZwcznMRPUtWvlo80CDAgDb1Vm9a6gY+qvIR2Aju5mg9e0tFEd8kSVT9Q634dPHtD/KEy/bu1RIoM4eHKc+9VKGWtaL0md0bk62ULsn/X97l5dMQvkURWTgicWaMhgW4BX4IKMOrGmGCb20ssOOAJf3eFvdwf0k78sUF8YGVAXSqDeuBuEaf3SB87FQd7ZGsnEJTWRdVAuRwGhduC7+QiR3d7kjk95YLcpNDdZUbE8dq5JvAUX5FIN/N9GdMF4A=
 
 before_script:
+    # to fasten up installation with parallel downloading
+    - composer global require hirak/prestissimo
```
Comment (TomasVotruba) [inline .travis.yml]: Awesome!

Do you know the numbers before/after?

File: .travis.yml [hunk 16]
```diff
@@ -21,14 +21,18 @@ env:
...
     # install composer dependencies
-    - composer update
+    - composer install
 
 script:
     # run tests for tested posts (with "tested: true" in header configuration)
     - vendor/bin/phpunit $PHPUNIT_FLAGS
     # check coding standard (defined in composer.json "scripts" section)
-    - if [[ "$CHECK_CS" != "" ]]; then composer check-cs; fi
+    - if [[ $CHECK_CS = 1 ]]; then composer check-cs; fi
```
Comment (TomasVotruba) [inline .travis.yml]: Nice cleanup

File: src/Posts/Year2017/SymfonyValidatorDynamicConstraints/Address.php [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+     * @return string
+     */
+    public function getZipcode(): string
+    {
+        return $this->zipcode;
+    }
+
+    /**
+     * @param string $zipcode
+     */
```
Comment (TomasVotruba) [inline src/Posts/Year2017/SymfonyValidatorDynamicConstraints/Address.php]: What are these docs for? It's already in code itself

File: tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php (hunk 10 above)
Comment (TomasVotruba) [inline tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php]: I know it makes sense - the rule itself. Just not in this case, the payload is already handled in in `run()` method:

```php
$this->payload = $this->payload ?: new \stdClass;
```

File: tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php [hunk 18]
```diff
@@ -87,22 +91,30 @@ public function testAddToBasketThirdProductRequest()
...
     {
+        /** @var Presenter $categoryPresenter */
         $categoryPresenter = $this->presenterFactory->createPresenter(self::PRESENTER_NAME);
         $categoryPresenter->autoCanonicalize = false;
 
         return $categoryPresenter;
     }
+
+
+    private function loadFileWithUnixLineEndings(string $file): string
```
Comment (TomasVotruba) [inline tests/Posts/Year2017/ListeningNetteComponents/Tests/ListeningNetteComponentsTest.php]: Nice naming!

File: tests/Posts/Year2017/SymfonyValidatorComparisonConstraints/ComparisonConstraintsTest.php [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+namespace Pehapkari\Website\Tests\Posts\Year2017\SymfonyValidatorComparisonConstraints;
+
+use Pehapkari\Website\Posts\Year2017\SymfonyValidatorComparisonConstraints\Event;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Validator\ConstraintViolationInterface;
+use Symfony\Component\Validator\ConstraintViolationListInterface;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+use Symfony\Component\Validator\ValidatorBuilder;
+
+class ComparisonConstraintsTest extends TestCase
```
Comment (TomasVotruba) [inline tests/Posts/Year2017/SymfonyValidatorComparisonConstraints/ComparisonConstraintsTest.php]: Rather `final` in here.

File: .travis.yml (hunk 15 above)
Comment (enumag) [inline .travis.yml]: no I didn't test that

File: .travis.yml (hunk 16 above)
Comment (enumag) [inline .travis.yml]: just copypasted from symplify :-D

File: src/Posts/Year2017/SymfonyValidatorDynamicConstraints/Address.php (hunk 17 above)
Comment (enumag) [inline src/Posts/Year2017/SymfonyValidatorDynamicConstraints/Address.php]: alright I'll clean that up

File: tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php (hunk 10 above)
Comment (enumag) [inline tests/Posts/Year2017/ListeningNetteComponents/Presenter/CategoryPresenter.php]: ok, I'll remove the call and add this to ignored errors

**Comment by enumag** [comment]
@TomasVotruba Thanks, all done. :-)

**Comment by TomasVotruba** [comment]
It it actually good you missed it :D ...

I found this test depends on extra php extension. That would make adding and article locally much more difficult. 

image


Could you make tests without that dependency or it necessary to demonstrate Validator usage?

**Comment by enumag** [comment]
The package contains the Zipcode constraint which is essential to my example use-case. And it's not something I can easily extract and copy here either.

On the other hand this particular constraint probably doesn't need BCmath so if you run composer with `--ignore-platform-reqs` it should work.

**Comment by TomasVotruba** [comment]
I understand the dependencies. But is it only example to demonstrate Validator features?
If you really need a ZIP service, could you use local dummy class which would and would not require and extension to do that?

I imagine a situation, when somebody would like to add an article for the firs time.

- clones the repository
- run composer update
- see the error abov
- will have learn how to install extension
- will have to find out which php.ini is located where, on Windows or Docker etc.
- and closes with exhaustion after 20 minutes

If this would an article about that zip package (why a ZIP package requires math extension putting aside) li), we can prevent that.

**Comment by TomasVotruba** [comment]
I found alternative: ZipCodeValidator, https://github.com/barbieswimcrew/zip-code-validator/blob/master/src/ZipCodeValidator/Constraints/ZipCodeValidator.php

Also it would make sense so implement this manually. E.g. we want to limit only to EU ZIPs that are stored in database etc. And test could be good source to explore and learn from, after reading the article.

**Comment by enumag** [comment]
@TomasVotruba It has nothing to do with ZIP, it's about Zipcode (= Postal code = PSČ). The validation needs [this](https://github.com/Soullivaneuh/IsoCodesValidator/blob/3.x/src/Constraints/ZipCodeValidator.php) and [this](https://github.com/ronanguilloux/IsoCodes/blob/master/src/IsoCodes/ZipCode.php) to work.

I understand your concerns but the same problem might occur with future articles as well. Using dummies everywhere might not always be possible (even here I don't know how to it) so it'a bad approach. The correct solution in my opinion would be for each tested article to have its local composer.json with it's dependencies completely separate from the rest.

Oh and I'm pretty sure that the alternative ZipCodeValidator you found is not compatible with the new Validator API intoroduces in Symfony 2.7 or so which means it's incompatible with Symfony 3.

**Comment by enumag** [comment]
@TomasVotruba Just imagine someone would write an article about a library that depends on the [PCNTL](http://php.net/manual/en/book.pcntl.php). There would be no way to avoid requiring the extension for the article tests. However there is no way to install that particular extension on Windows simply because it's not implemented. Then what? Disallowing tests for such article?

**Comment by TomasVotruba** [comment]
I meant ZIP in ZipCode.

You write about Validator, not math library. There is no added value in increasing complexity.

**Comment by enumag** [comment]
Well the low complexity solution doesn't seem to work here. :)

**Comment by enumag** [comment]
@TomasVotruba Alright so here is another idea:

- remove the library from composer.json
- exclude the test by default using phpunit.xml
- use a different phpunit.xml when running tests on travis
- install the library on travis only so that the test may pass

**Comment by TomasVotruba** [comment]
Or use single class with own method.
It would also show how to make a constraint.

The test violates basic principles of unit testing. It's like testing entity creation by installing whole Doctrine.

Test a unit, not unused package

**Comment by TomasVotruba** [comment]
That Validator magic gave me a hell :D uff.

All final changes are here: https://github.comhis/pehapkari/pehapkari.cz/commit/5709ccaccdfd91c62f0ea753ff38e5f2661c6970

Articles will be published (in rss and blog page) in those dates... They are available in their urls:

- https://pehapkari.cz/blog/2017/02/11/symfony-validator-comparison-constraints/
- https://pehapkari.cz/blog/2017/02/18/symfony-validator-conditional-constraints/
- https://pehapkari.cz/blog/2017/02/18/symfony-validator-conditional-constraints/

You can still send fixes, if you like.

(Please schedule Tweet on the day of publishing, not earlier to avoid confusions).


I think it could be simpler, but I'm tired from all those counter-spells :D 

**You've been awarded 30 points to the contest!**
https://github.com/pehapkari/pehapkari.cz/issues/178

**Comment by enumag** [comment]
@TomasVotruba You surely made a lot of changes. :-) Thanks! I see that my articles are not listed yet. But sometimes there are areticles with future days listed on the web (there is one now as well). So how does it work?

**Comment by TomasVotruba** [comment]
@enumag That's the last one. Others will be seen when date is right. See:

https://github.com/pehapkari/pehapkari.cz/commit/3d6c7ce42a9f3744e51213343ed760b68b9ffed5

**Comment by enumag** [comment]
@TomasVotruba Thanks for clarification. Isn't the comparison a bit wrong though? https://github.com/pehapkari/pehapkari.cz/commit/3d6c7ce42a9f3744e51213343ed760b68b9ffed5#commitcomment-20654922

**Comment by TomasVotruba** [comment]
@enumag Just FYI: I get notification about both comments, so no need to reference it twice.
